### PR TITLE
Dns policy

### PR DESCRIPTION
**Pull request by ls-2018**
# Description
If DNSPolicy is not ClusterFirst, the dapr sidecar may not able to connect to other dapr sidecar.
 
## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by ls-2018** [comment]
It's not the same question. I'm going to give you an example to prove that the change is correct.
I have two programs `foo`, `bar`, where `foo` calls `bar`, <s>but it seems that the logic is not used.</s>
As you can see, when we set the pod's dns policy to Default, the daprd container will never be ready, let alone the routing behind it.There is nothing wrong with the logic of the program itself.In this case, it should not be injected.

### bar.py
```
import time

from flask import Flask

app = Flask("___")


@app.route('/')
@app.route('/index')
def index():
    return "bar:" + str(time.time())


if __name__ == '__main__':
    app.run(port=8000)
```
### foo.py
```
import requests

from flask import Flask

app = Flask("___")


@app.route('/')
def index():
    try:
        return requests.get(f"http://localhost:3500/v1.0/invoke/bar/method/index").text
    except Exception as e:
        return str(e)


if __name__ == '__main__':
    app.run(port=8000)

```

### pod.yaml
```
---
kind: Deployment
apiVersion: apps/v1
metadata:
  name: foo
spec:
  template:
    metadata:
      annotations:
        dapr.io/enabled: 'true'
        dapr.io/app-id: foo
        dapr.io/app-port: '8000'
      labels:
        app: foo
    spec:
      dnsPolicy: "Default"
      containers:
        - name: foo
          image: acejilam/example:latest
          args:
            - "/app/foo.py"
  selector:
    matchLabels:
      app: foo
---
kind: Deployment
apiVersion: apps/v1
metadata:
  name: bar
spec:
  template:
    metadata:
      annotations:
        dapr.io/enabled: 'true'
        dapr.io/app-id: bar
        dapr.io/app-port: '8000'
      labels:
        app: bar
    spec:
      containers:
        - name: bar
          image: acejilam/example:latest
          args:
            - "/app/bar.py"
  selector:
    matchLabels:
      app: bar

```

### Dockerfile
```
#FROM python:3.9.6-slim as base
FROM centos:8 as base
RUN #sed -i 's#http://deb.debian.org#https://mirrors.aliyun.com#g' /etc/apt/sources.list
RUN #sed -i 's#http://security.debian.org#https://mirrors.aliyun.com#g' /etc/apt/sources.list
RUN rm -f /etc/localtime && ln -sv /usr/share/zoneinfo/Asia/Shanghai /etc/localtime && echo "Asia/Shanghai" > /etc/timezone
RUN yum install python3.9 epel-release  lsof -y && yum install htop -y
RUN #apt-get update && apt-get install vim -y
#mirrors.aliyun.com|mirror.lzu.edu.cn|mirrors.ustc.edu.cn|mirrors.cloud.tencent.com
# ---- compile image -----------------------------------------------
FROM base AS compile-image
RUN python3 -m pip install --upgrade pip
RUN python3 -m venv /venv
RUN #apt-get update -y && apt-get install -y build-essential
# pip install is fast here (while slow without the venv) :
RUN pip3 config set global.index-url https://pypi.tuna.tsinghua.edu.cn/simple
RUN /venv/bin/python3 -m pip install --upgrade pip
RUN /venv/bin/pip3 install flask requests
# ---- build image -----------------------------------------------
FROM base AS build-image
COPY --from=compile-image /venv /venv
## Make sure we use the virtualenv:
COPY . /app
WORKDIR /app
ENV PATH="/venv/bin:$PATH"
#RUN . /venv/bin/activate
ENV LC_ALL zh_CN.UTF-8
ENV LANG zh_CN.UTF-8
EXPOSE 8000
ENTRYPOINT ["/venv/bin/python3"]

```

### test.sh
```
docker build -t acejilam/example:latest .
docker push acejilam/example:latest
kind delete cluster
echo -e 'kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  extraPortMappings:
  - containerPort: 80
    hostPort: 80
    listenAddress: "0.0.0.0" # Optional, defaults to "0.0.0.0"
    protocol: udp # Optional, defaults to tcp
- role: worker
' > ./kind.yaml

kind create cluster --config ./kind.yaml
kubectl cluster-info --context kind-kind
dapr init -k
kubectl apply -f pod.yaml
sleep 300
# shellcheck disable=SC2046
kubectl logs $(kubectl get pod|grep foo|awk '{print $1}') -c daprd
#kubectl exec -it $(kubectl get pod|grep Running|grep foo|awk '{print $1}') -c foo -- curl http://127.0.0.1:8000/
```
image
image

File: pkg/injector/injector.go [hunk 1]
```diff
@@ -224,6 +234,7 @@ func (i *injector) handleRequest(w http.ResponseWriter, r *http.Request) {
 		}
 	}
 
+SkipPatch:
```
Comment (daixiang0) [inline pkg/injector/injector.go:237]: I really do not like any `goto` codes, could we use other way to do it?

**Comment by daixiang0** [comment]
If we add the DNS component mentioned in #4065, the policy-related checks should be in that component.

If not, we need this check to ensure Dapr sidecar works.

/cc @artursouza

**Comment by ls-2018** [comment]
I then add some data to illustrate the difference between the #4264  #4065

#4264：

image
image
image
 
We can see that foo Container based on 'dnsPolicy: "Default" configuration cannot resolve the k8S internal service name. Because he hasn't been resolved by CoreDNS.
In the case of 'DirectMessage', there is communication through the K8S service name, but in this case there is no communication at all. Other components based on K8S domain name communication will be affected.
image


#4065:
This proposal is intended to support communication between different components using different name resolution components.
For example, Foo and Bar communicate with each other using Consul.
DNS communication is used between foo2 and bar2.
However, no matter which parsing component is used, a service is required to implement this functionality.
If DAPRD and the service communicate via IP :port, then this case is useless.
However, if it is resolved by domain name, how can it be implemented in K8S cluster without coreDNS?【The external DNS of the cluster is excluded】

File: pkg/injector/injector.go (hunk 1 above)
Comment (ls-2018) [reply pkg/injector/injector.go:237]: I'll change it.

File: pkg/injector/injector.go [hunk 2]
```diff
@@ -209,9 +209,21 @@ func (i *injector) handleRequest(w http.ResponseWriter, r *http.Request) {
...
 	ar := v1.AdmissionReview{}
 	_, gvk, err := i.deserializer.Decode(body, nil, &ar)
+	isDNSClusterFirst := true
 	if err != nil {
 		log.Errorf("Can't decode body: %v", err)
 	} else {
+		var pod corev1.Pod
+		if err = json.Unmarshal(ar.Request.Object.Raw, &pod); err != nil {
+			log.Warnf("could not unmarshal raw object: %v", err)
+		} else {
```
Comment (daixiang0) [inline pkg/injector/injector.go:219]: Codes look good, just a nit: use `elsif` here.

Comment (ls-2018) [reply pkg/injector/injector.go]: 👌🏻

**Comment by yaron2** [comment]
@ls-2018 please fix the failing checks.

**Comment by ls-2018** [comment]
golangci-lint run -c .golangci.yml --fix

**Pull request closed without merging**